lighthouse-core/gather/driver.js

```diff
@@ -688,6 +688,20 @@ class Driver {
   }
 
   /**
+   * Returns the flattened list of all DOM nodes within the document.
+   * @param {boolean=} pierce Whether to pierce through shadow trees and iframes.
+   *     True by default.
+   * @return {!Promise<!Array<!Element>>} The found elements, or [], resolved in a promise
+   */
+  getAllNodesInDocument(pierce = true) {
```

Comment:
@paulirish @ebidel this is what I had come up with for querying through shadow dom.

```js
  queryAllElements(query, includeUserAgentShadowDOM = true) {
    return this.sendCommand('DOM.performSearch', {query, includeUserAgentShadowDOM})
      .then(({searchId, resultCount}) => {
        if (resultCount === 0) {
          return [];
        }

        return this.sendCommand('DOM.getSearchResults', {
          searchId,
          fromIndex: 0,
          toIndex: resultCount
        });
      })
      .then(nodeList => {
        const elementList = [];
        nodeList.nodeIds.forEach(nodeId => {
          if (nodeId !== 0) {
            elementList.push(new Element({nodeId}, this));
          }
        });
        return elementList;
      });
  }
```